Add docs (#13)
huggingface/hffs: docs/source/index.mdx

# Filesystem

🤗 Filesystem is a package that provides a pythonic file interface ([fsspec-compatible](https://filesystem-spec.readthedocs.io/en/latest/)) to the [Hugging Face Hub](https://huggingface.co/). It builds on top of [Hugging Face Hub client library](https://huggingface.co/docs/huggingface_hub/index) to interact with the Hub.

## Installation

```bash
pip install hffs
```

## Usage

> `HfFileSystem` is the cental class of the library that holds connection information and allows typical filesystem style operations like `cp`, `mv`, `ls`, `du`, `glob`, `get_file`, `put_file` etc.

Reviewer: Depending on the audience you think will use 🤗 Filesystems, it might help beginners to briefly explain filesystem operations. It may also be nice to add a link where users can see all the supported filesystem operations so they can get a more complete understanding of 🤗 Filesystems capabilities. 

But if this is mostly aimed at more experienced developers or you think it's something users know already (things like `get_file` and `put_file` are mostly new to me), feel free to ignore!

```suggestion
`HfFileSystem` is the library's main class that holds connection information and enables typical filesystem style operations like `cp`, `mv`, `ls`, `du`, `glob`, `get_file`, `put_file` etc.
```

```python
>>> from hffs import HfFileSystem
>>> fs = HfFileSystem("my-username/my-dataset-repo", repo_type="dataset")

>>> # List files in a directory
>>> fs.ls("data")
['train.csv', 'test.csv']

>>> # List all ".csv" files in a repo
>>> fs.glob("**.csv")
['data/train.csv', 'data/test.csv']

>>> # Read the contents of a remote file 
>>> with fs.open("data/train.csv", "r") as f:
...     train_data = f.readlines()

>>> # Read all the contents of a remote file at once as a string
>>> train_data = fs.read_text("data/train.csv")

>>> # Write a remote file
>>> with fs.open("data/validation.csv", "w") as f:
...     f.write("text,label")
...     f.write("Fantastic movie!,good")
```

If not explicitly provided, the `repo_type` argument defaults to `model`. Besides `model`, other supported values are `dataset` and `space`.

> The optional `revision` argument can be passed to open a filesystem from a specific commit (any revision such as a branch or a tag name or a commit hash).

Reviewer: Maybe it'd be good to couple this with one of the code snippets above so users can see the code and comment :)

> Note that, unlike the built-in `open`, the mode in `fsspec`'s `open` defaults to binary mode, i.e. "rb". This means you must explicitly set encoding as "r"/"w" in case of text mode.

Reviewer: ```suggestion
Unlike Python's built-in `open`, `fsspec`'s `open` defaults to binary mode, `"rb"`. This means you must explicitly set encoding as `"rw"` for reading and writing in text mode.
```

> ## Integration

Reviewer: Maybe it'd make more sense to move the Integration section before the Usage section? It might be good for the user to check if they can use a URL with an integration before they start using the filesystem operations.

Author: This order comes from the [`s3fs`](https://s3fs.readthedocs.io/en/latest/) docs, so I think I'll leave it as-is.

> To use 🤗 Filesystem with a library that accepts `fsspec`'s URLs, the URL must be of the following structure:

Reviewer: ```suggestion
🤗 Filesystem can be used with any library that integrates `fsspec`, and the URL has the following structure:
```

```
hf://[<repo_type>/]<repo_id>[@<revision>]:/<path/in/repo>
```

> The `repo_type` and `revision` parameters are optional. Most integrations also make it possible to pass optional parameters to the filesystem's initializer as `storage_options`, a dictionary mapping parameter names to their values.

Reviewer: ```suggestion
The `repo_type` and `revision` parameters are optional. Most integrations also allow you to pass optional parameters to the filesystem's initializer as `storage_options`, a dictionary mapping parameter names to their values.
```

> Before passing a URL to integration, you can use `fsspec.get_fs_token_paths` to ensure the URL initializes a filesystem as expected:

Reviewer: ```suggestion
Before passing a URL to an integration, you can use `fsspec.get_fs_token_paths` to ensure the URL initializes a filesystem as expected:
```

```python
>>> import fsspec

>>> def get_filesystem_state_and_path_from_url(url):
...     fs, _, paths = fsspec.get_fs_token_paths(url)
...     return {"repo_type": fs.repo_type, "repo_id": fs.repo_id, "revision": fs.revision or "main", "path/in/repo": paths[0]}

>>> print(get_filesystem_state_and_path_from_url("hf://dataset/johndoe/cityskapes:/images/image0.jpg"))
{'repo_type': 'dataset', 'repo_id': 'johndoe/cityskapes', 'revision': 'main', 'path/in/repo': 'images/image0.jpg'}

>>> print(get_filesystem_state_and_path_from_url("hf://johndoe/cityskapes@dev:/"))
{'repo_type': 'model', 'repo_id': 'johndoe/cityskapes', 'revision': 'dev', 'path/in/repo': ''}
```

> ## Authentication

Reviewer: Maybe also move this under Integration so users know beforehand they should be logged in with their Hugging Face account. So:

```
## Integration
## Authentication
## Usage
```

> In many cases, you must be logged in with a Hugging Face account to interact with the Hub. Refer to the [Login](https://huggingface.co/docs/huggingface_hub/quick-start#login) section of the Hugging Face Hub client library documentation to learn about the ways to authenticate on the Hub. It is also possible to log in programmatically by passing your `token` as an argument to `HfFileSystem`. If you do so, be careful when
> sharing your source code to avoid accidentally leaking the token.

Reviewer: ```suggestion
In many cases, you must be logged in with a Hugging Face account to interact with the Hub. Refer to the [Login](https://huggingface.co/docs/huggingface_hub/quick-start#login) section of the Hugging Face Hub client library documentation to learn about authentication methods on the Hub. 

It is also possible to login programmatically by passing your `token` as an argument to `HfFileSystem`. If you login this way, be careful not to accidentally leak the token when sharing your source code!
```

Before passing a URL to an integration, you can use `fsspec.get_fs_token_paths` to ensure the URL initializes a filesystem as expected:

```python
>>> import fsspec

>>> def get_filesystem_state_and_path_from_url(url):
...     fs, _, paths = fsspec.get_fs_token_paths(url)
...     return {"repo_type": fs.repo_type, "repo_id": fs.repo_id, "revision": fs.revision or "main", "path/in/repo": paths[0]}

>>> print(get_filesystem_state_and_path_from_url("hf://dataset/johndoe/cityskapes:/images/image0.jpg"))
{'repo_type': 'dataset', 'repo_id': 'johndoe/cityskapes', 'revision': 'main', 'path/in/repo': 'images/image0.jpg'}

>>> print(get_filesystem_state_and_path_from_url("hf://johndoe/cityskapes@dev:/"))
{'repo_type': 'model', 'repo_id': 'johndoe/cityskapes', 'revision': 'dev', 'path/in/repo': ''}
```

## Authentication

In many cases, you must be logged in with a Hugging Face account to interact with the Hub. Refer to the [Login](https://huggingface.co/docs/huggingface_hub/quick-start#login) section of the Hugging Face Hub client library documentation to learn about authentication methods on the Hub. 

> It is also possible to login programmatically by passing your `token` as an argument to `HfFileSystem`. If you login this way, be careful not to accidentally leak the token when sharing your source code!

Reviewer: Maybe a code example here showing how to pass your `token` to `HfFileSystem`?
